libraries/psio/include/psio/fracpack.hpp

```diff
@@ -0,0 +1,2608 @@
+      size_t size() const
+      {
+         if (_data == nullptr)
+            return 0;
+         uint32_t s;
+         memcpy(&s, _data.get(), sizeof(s));
+         return s;
+      }
+
+      void reset() { _data.reset(); }
+
+      void resize(uint32_t size)
+      {
+         _data.reset();
+         _data = std::shared_ptr<char>(new char[size + sizeof(size)], [](char* c) { delete[] c; });
+         memcpy(_data.get(), &size, sizeof(size));
+      }
+
+      bool validate(bool& unknown) const
+      {
+         if (_data)
+         {
+            auto r  = psio::fracvalidate<T>(data(), data() + size());
+            unknown = r.unknown;
+            return r.valid;
+         }
+         return false;
+      }
+      bool validate() const
+      {
+         if (_data)
+         {
+            return psio::fracvalidate<T>(data(), data() + size()).valid;
+         }
+         return false;
+      }
+      bool validate_all_known() const
+      {
+         if (_data)
+         {
+            return psio::fracvalidate<T>(data(), data() + size()).valid_and_known();
+         }
+         return false;
+      }
+
+      T unpack() const
+      {
+         T            tmp{};
+         input_stream in(_data.get() + sizeof(uint32_t), size());
+         psio::fracunpack(tmp, in);
+         return tmp;
+      }
+
+     private:
+      std::shared_ptr<char> _data;
+   };
+   template <>
+   class shared_view_ptr<void>
+   {
+   };
+
+   template <typename T, typename S>
+   void to_json(const shared_view_ptr<T>& obj, S& stream)
+   {
+      to_json(obj.unpack(), stream);
+   }
+
+   template <typename T, typename S>
+   void from_json(shared_view_ptr<T>& obj, S& stream)
+   {
+      if constexpr (std::is_same_v<T, std::string>)
+         // ambiguous case
+         shared_view_ptr<T>::from_json_undefined();
+      if (stream.peek_token().get().type == json_token_type::type_string)
+      {
+         // TODO: avoid copy
+         std::vector<char> v;
+         from_json(v, stream);
+         obj = {v.data(), v.size()};
+      }
+      else
+      {
+         T inner;
+         from_json(inner, stream);
+         obj = inner;
+      }
+   }
+
+   template <typename T, typename R, typename... Args>
+   constexpr auto tuple_from_function_args(R (T::*func)(Args...))
+       -> std::tuple<std::decay_t<Args>...>;
+
+   template <typename T, typename R, typename... Args>
+   constexpr auto tuple_from_function_args(R (T::*func)(Args...) const)
+       -> std::tuple<std::decay_t<Args>...>;
+
+   template <typename... Args>
+   constexpr auto tuple_remove_view(std::tuple<Args...>) -> std::tuple<remove_view_t<Args>...>;
+
+   template <typename T>
+   std::vector<char> to_frac(const T& v)
+   {
+      std::vector<char> result(fracpack_size(v));
+      fast_buf_stream   s(result.data(), result.size());
+      fracpack(v, s);
+      return result;
+   }
+
+   template <typename T>
+   T from_frac(const std::vector<char>& b)
```

Comment:
It isn't obvious to me why `from_frac` means checked while `convert_from_frac` means unchecked.